src/lib/storage/frame_of_reference_segment/frame_of_reference_segment_iterable.hpp

```diff
@@ -56,79 +55,50 @@ class FrameOfReferenceSegmentIterable : public PointAccessibleSegmentIterable<Fr
+                      OffsetValueDecompressorT attribute_decompressor, ChunkOffset chunk_offset)
+        : _block_minima{block_minima},
+          _null_values{null_values},
+          _offset_value_decompressor{std::move(attribute_decompressor)},
           _chunk_offset{chunk_offset} {}
 
    private:
     friend class boost::iterator_core_access;  // grants the boost::iterator_facade access to the private interface
 
-    void increment() {
-      ++_offset_value_it;
-      ++_null_value_it;
-      ++_index_within_frame;
-      ++_chunk_offset;
+    void increment() { ++_chunk_offset; }
 
-      if (_index_within_frame >= FrameOfReferenceSegment<T>::block_size) {
-        _index_within_frame = 0u;
-        ++_block_minimum_it;
-      }
-    }
+    void decrement() { --_chunk_offset; }
 
-    void decrement() {
-      --_offset_value_it;
-      --_null_value_it;
-      --_chunk_offset;
-
-      if (_index_within_frame > 0) {
-        --_index_within_frame;
-      } else {
-        _index_within_frame = FrameOfReferenceSegment<T>::block_size - 1;
-        --_block_minimum_it;
-      }
-    }
+    void advance(std::ptrdiff_t n) { _chunk_offset += n; }
+
+    bool equal(const Iterator& other) const { return _chunk_offset == other._chunk_offset; }
 
-    void advance(std::ptrdiff_t n) {
-      // For now, the lazy approach
-      PerformanceWarning("Using repeated increment/decrement for random access");
-      if (n < 0) {
-        for (std::ptrdiff_t i = n; i < 0; ++i) {
-          decrement();
-        }
-      } else {
-        for (std::ptrdiff_t i = 0; i < n; ++i) {
-          increment();
-        }
-      }
+    std::ptrdiff_t distance_to(const Iterator& other) const {
+      return static_cast<std::ptrdiff_t>(other._chunk_offset) - _chunk_offset;
     }
 
-    bool equal(const Iterator& other) const { return _offset_value_it == other._offset_value_it; }
+    SegmentPosition<T> dereference() const {
+      static constexpr auto block_size = FrameOfReferenceSegment<T>::block_size;
 
-    std::ptrdiff_t distance_to(const Iterator& other) const { return other._offset_value_it - _offset_value_it; }
+      const auto is_null = (*_null_values)[_chunk_offset];
+      const auto block_minimum = (*_block_minima)[_chunk_offset / block_size];
+      const auto offset_value = _offset_value_decompressor.get(_chunk_offset);
+      const auto value = static_cast<T>(offset_value) + block_minimum;
 
-    SegmentPosition<T> dereference() const {
-      const auto value = static_cast<T>(*_offset_value_it) + *_block_minimum_it;
-      return SegmentPosition<T>{value, *_null_value_it, _chunk_offset};
+      return SegmentPosition<T>{value, is_null, _chunk_offset};
     }
 
    private:
-    ReferenceFrameIterator _block_minimum_it;
-    OffsetValueIteratorT _offset_value_it;
-    NullValueIterator _null_value_it;
-    size_t _index_within_frame;
+    const pmr_vector<T>* _block_minima;
+    const pmr_vector<bool>* _null_values;
```

Comment:
Why is this a pointer, not a reference?

Reply:
Because iterators need to be copyable. And we did it similar in other iterators before. 😁
Am happy to consider alternatives.

Reply:
Shouldn't that be possible with references, too? Unless I'm mistaken, references are only a problem if you want to assign the class (which is not possible right now either because the pointer is const)

Reply:
Afaik, the copy assignment does not work with references. At that's what is being done a lot with iterators. What might work is using a `std::reference_wrapper`.

Reply:
Let me have a look at this tomorrow or so

Reply:
Benchmark is running...

Reply:
Ok, we won't get around having something assignable here. reference_wrappers are not really better just because they are not a pointer.

Reply:
I actually kind of like them, because they "just work" and they cost have a cost here. But when you and Herb Sutter (https://herbsutter.com/2020/02/) don't like them, ok.